### PR TITLE
EffBreakNaturally

### PR DESCRIPTION
**Pull request by ShaneBeee**
### Description
New effect for breaking blocks naturally, with an option to add a tool (This will break the block based on how said tool would, ie: a diamond pickaxe of fortune 3 would drop more coal items when breaking coal ore blocks)

---
**Target Minecraft Versions:**  Any version
**Requirements:** none
**Related Issues:** none

### CODE REVIEW

**Comment by ShaneBeee** [comment]
> > The method from spigot calls for an ItemStack rather than a type, and because of that I was running into problems with it not working.
> > If it's completely necessary I can try to change it again.
> 
> You use `getRandom()` method from an ItemType, the name is weird but it's old, it converts an ItemType into an ItemStack based off the ItemType properties.

OOOOO good to know, thanks! 
Would that also work for something like:
```
break block naturally using player's tool
```
??

**Comment by TheLimeGlass** [comment]
> 
> 
> > > The method from spigot calls for an ItemStack rather than a type, and because of that I was running into problems with it not working.
> > > If it's completely necessary I can try to change it again.
> > 
> > 
> > You use `getRandom()` method from an ItemType, the name is weird but it's old, it converts an ItemType into an ItemStack based off the ItemType properties.
> 
> OOOOO good to know, thanks!
> Would that also work for something like:
> 
> ```
> break block naturally using player's tool
> ```
> 
> ??

Yes, it is recommended to always use ItemType when handling any items in Skript. ItemType takes in an ItemType or ItemStack.

It makes this following syntax possible aswell
```
break target block naturally with a diamond pickaxe
```

**Comment by ShaneBeee** [comment]
> > > > The method from spigot calls for an ItemStack rather than a type, and because of that I was running into problems with it not working.
> > > > If it's completely necessary I can try to change it again.
> > > 
> > > 
> > > You use `getRandom()` method from an ItemType, the name is weird but it's old, it converts an ItemType into an ItemStack based off the ItemType properties.
> > 
> > 
> > OOOOO good to know, thanks!
> > Would that also work for something like:
> > ```
> > break block naturally using player's tool
> > ```
> > ??
> 
> Yes, it is recommended to always use ItemType when handling any items in Skript. ItemType takes in an ItemType or ItemStack.
> 
> It makes this following syntax possible aswell
> 
> ```
> break target block naturally with a diamond pickaxe
> ```

Well that is very good to know.... thanks, I appreciate the help, gonna get on fixing that now!

**Comment by TheLimeGlass** [comment]
furtune is spelt wrong in the examples.

File: src/main/java/ch/njol/skript/effects/EffBreakNaturally.java [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+	public boolean init(final Expression<?>[] exprs, final int matchedPattern, final Kleenean isDelayed, final SkriptParser.ParseResult parser) {
+		block = (Expression<Block>) exprs[0];
+		tool = (Expression<ItemStack>) exprs[1];
+		return true;
+	}
+	
+	@Override
+	protected void execute(final Event e) {
+		ItemStack tool = this.tool != null ? this.tool.getSingle(e) : null;
+		for(Block block : this.block.getArray(e)){
```
Comment (TheLimeGlass) [inline src/main/java/ch/njol/skript/effects/EffBreakNaturally.java]: Also please add spaces in blocks `for (Block block : this.block.getArray(e)) {` This is a standard in Skript and majority of Java. Similar on the next line in the if condition.

**Comment by ShaneBeee** [comment]
> furtune is spelt wrong in the examples.

its the canadian spelling?!?! 😜

File: src/main/java/ch/njol/skript/effects/EffBreakNaturally.java (hunk 1 above)
Comment (ShaneBeee) [reply src/main/java/ch/njol/skript/effects/EffBreakNaturally.java]: Ahh sorry... 
I took over a plugin a while back, and thats how the dev did it, guess it got stuck in my head!

File: src/main/java/ch/njol/skript/effects/EffBreakNaturally.java [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+import ch.njol.util.Kleenean;
+
+@Name("Break Block Naturally")
+@Description({"Breaks the block and spawns items as if a player had mined it",
+		"\nYou can add a tool, which will spawn items based on how that tool would break the block ",
+		"(ie: Using a diamond pickaxe of fortune 3 would drop more coal items when breaking a coal ore block)"})
+@Examples({"on right click:", "\tbreak clicked block naturally",
+		"loop blocks in radius 10 around player:", "\tbreak loop-block naturally using player's tool",
+		"loop blocks in radius 10 around player:", "\tbreak loop-block naturally using diamond pickaxe of fortune 3"})
+@Since("{INSERT VERSION}")
```
Comment (Nicofisi) [inline src/main/java/ch/njol/skript/effects/EffBreakNaturally.java]:
```suggestion
@Since("INSERT VERSION")
```
Wasn't it just that? :eyes:

Comment (ShaneBeee) [reply src/main/java/ch/njol/skript/effects/EffBreakNaturally.java]: There's a high chance you are right. 
Im not actually sure why I thought the brackets were there ¯\_(ツ)_/¯ 
I probably saw it somewhere, but honestly, I have no clue

**Comment by Snow-Pyon** [comment]
> You use `getRandom()` method from an ItemType, the name is weird but it's old, it converts an ItemType into an ItemStack based off the ItemType properties.

Just clarifying, this isn't actually the case. The method is named that way because that's literally what it does, get a random ItemData from the ItemType and return it as ItemStack. 

People often tend to think that ItemType is just something like ItemStack but it isn't exactly the case. ItemType is a holder for one or more ItemDatas, ItemDatas being the specific items (e.g: `diamond pickaxe` or `gold ore`) and ItemType being a group of those (e.g: `any pickaxe` or `any ore`), it is just that it often holds a single ItemData, which is why the getRandom method returns that, but if you use it in ItemTypes like `any ore`, it will actually choose a random ItemData from that group.

File: src/main/java/ch/njol/skript/effects/EffBreakNaturally.java [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+		block = (Expression<Block>) exprs[0];
+		tool = (Expression<ItemType>) exprs[1];
+		return true;
+	}
+	
+	@Override
+	protected void execute(final Event e) {
+		ItemType tool = this.tool != null ? this.tool.getSingle(e) : null;
+		for (Block block : this.block.getArray(e)){
+			if (tool != null) block.breakNaturally(tool.getRandom());
```
Comment (Pikachu920) [inline src/main/java/ch/njol/skript/effects/EffBreakNaturally.java]: have you tried without the conditions? calling it with null probably does the same thing as the no arg call

File: src/main/java/ch/njol/skript/effects/EffBreakNaturally.java [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+		"loop blocks in radius 10 around player:", "\tbreak loop-block naturally using diamond pickaxe of fortune 3"})
+@Since("INSERT VERSION")
+public class EffBreakNaturally extends Effect {
+	
+	static {
+		Skript.registerEffect(EffBreakNaturally.class, "break %blocks% naturally [using %-itemtype%]");
+	}
+	
+	@SuppressWarnings("null")
+	private Expression<Block> block;
```
Comment (Pikachu920) [inline src/main/java/ch/njol/skript/effects/EffBreakNaturally.java]: should be plural (blocks)

File: src/main/java/ch/njol/skript/effects/EffBreakNaturally.java [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+		"\nYou can add a tool, which will spawn items based on how that tool would break the block ",
+		"(ie: Using a diamond pickaxe of fortune 3 would drop more coal items when breaking a coal ore block)"})
+@Examples({"on right click:", "\tbreak clicked block naturally",
+		"loop blocks in radius 10 around player:", "\tbreak loop-block naturally using player's tool",
+		"loop blocks in radius 10 around player:", "\tbreak loop-block naturally using diamond pickaxe of fortune 3"})
+@Since("INSERT VERSION")
+public class EffBreakNaturally extends Effect {
+	
+	static {
+		Skript.registerEffect(EffBreakNaturally.class, "break %blocks% naturally [using %-itemtype%]");
```
Comment (Pikachu920) [inline src/main/java/ch/njol/skript/effects/EffBreakNaturally.java]: naturally should be optional

File: src/main/java/ch/njol/skript/effects/EffBreakNaturally.java (hunk 4 above)
Comment (ShaneBeee) [reply src/main/java/ch/njol/skript/effects/EffBreakNaturally.java]: odd thing to be picky about but ok

File: src/main/java/ch/njol/skript/effects/EffBreakNaturally.java (hunk 5 above)
Comment (ShaneBeee) [reply src/main/java/ch/njol/skript/effects/EffBreakNaturally.java]: Should be, or is this your opinion?
I don't mean to sound rude here, but I have no idea why naturally "should" be optional!
The Bukkit method is breakNaturally() after all.

File: src/main/java/ch/njol/skript/effects/EffBreakNaturally.java (hunk 4 above)
Comment (Pikachu920) [reply src/main/java/ch/njol/skript/effects/EffBreakNaturally.java]: it may seem like it, but variable naming is a big deal

File: src/main/java/ch/njol/skript/effects/EffBreakNaturally.java (hunk 3 above)
Comment (ShaneBeee) [reply src/main/java/ch/njol/skript/effects/EffBreakNaturally.java]: So now Im confused as to which direction to go.
Pikachu says change it, Bensku says probably not?!?!?

File: src/main/java/ch/njol/skript/effects/EffBreakNaturally.java (hunk 4 above)
Comment (ShaneBeee) [reply src/main/java/ch/njol/skript/effects/EffBreakNaturally.java]: it doesn't appear that way, there's variables in another class (needles/haystack)  ;)

File: src/main/java/ch/njol/skript/effects/EffBreakNaturally.java (hunk 3 above)
Comment (Pikachu920) [reply src/main/java/ch/njol/skript/effects/EffBreakNaturally.java]: I'd trust bensku, I know he has spent time in therapy after dealing with NMS npes

Comment (ShaneBeee) [reply src/main/java/ch/njol/skript/effects/EffBreakNaturally.java]: Poor guy! Ill leave it as is then.

File: src/main/java/ch/njol/skript/effects/EffBreakNaturally.java (hunk 5 above)
Comment (TheLimeGlass) [reply src/main/java/ch/njol/skript/effects/EffBreakNaturally.java]: There is no syntax that mimics `break %blocks%` so adding this would be a good addition.

Comment (ShaneBeee) [reply src/main/java/ch/njol/skript/effects/EffBreakNaturally.java]: On that note.... (I will change this) but should the name be
```
@Name("Break Block Naturally")
```
or
```
@Name("Break Block")
```

**Comment by ShaneBeee** [comment]
okay @Pikachu920 this has all been updated to your suggestions, anything else you feel I need to change?

File: src/main/java/ch/njol/skript/effects/EffBreakNaturally.java [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+		ItemType tool = this.tool != null ? this.tool.getSingle(e) : null;
+		for (Block block : this.blocks.getArray(e)){
+			if (tool != null) block.breakNaturally(tool.getRandom());
+			else block.breakNaturally();
+		}
+	}
+	
+	@Override
+	public String toString(final @Nullable Event e, final boolean debug) {
+		return "break " + blocks.toString(e, debug) + " naturally" + (tool != null ? " using " + tool : "");
```
Comment (Pikachu920) [inline src/main/java/ch/njol/skript/effects/EffBreakNaturally.java]: toString is wrong for tool

Comment (ShaneBeee) [reply src/main/java/ch/njol/skript/effects/EffBreakNaturally.java]: OOPS.... I'm on it!

File: src/main/java/ch/njol/skript/effects/EffBreakNaturally.java (hunk 3 above)
Comment (Pikachu920) [reply src/main/java/ch/njol/skript/effects/EffBreakNaturally.java]: what about a case where tool is null not because the expression is null but because the value was. should that really still break the block?

Comment (Snow-Pyon) [reply src/main/java/ch/njol/skript/effects/EffBreakNaturally.java]: Yeah, you're still able to break a block if you're using air (which is practically null so)

**Comment by Pikachu920** [comment]
still wondering on the null tool behavior

**Comment by ShaneBeee** [comment]
> still wondering on the null tool behavior

I tested with this
```
on right click:
	break clicked block naturally using air
```
no issues

I also tested with this, once using a shovel and once using an empty hand
```
on right click:
	break clicked block naturally using player's tool
``` 
also no issues.

**Comment by Pikachu920** [comment]
neither of those are null lol,
I mean like with this code `break clicked block with {_null}` where {_null} is not set will still break the block even though {_null} didn't have a value. The question is: is that behavior confusing?

**Comment by TheBentoBox** [comment]
> I mean like with this code `break clicked block with {_null}` where {_null} is not set will still break the block even though {_null} didn't have a value. The question is: is that behavior confusing?

Personally, I'd expect that to work fine and break it as if your tool was air.

**Comment by Pikachu920** [comment]
in my mind it wouldn't break because o obvious intended it to break with *some* tool

**Comment by ShaneBeee** [comment]
Holy crap I feel slow, I finally understood what you mean.
Okay so I changed it to this
```java
	@Override
	@SuppressWarnings("null")
	protected void execute(final Event e) {
		ItemType tool = this.tool != null ? this.tool.getSingle(e) : null;
		for (Block block : this.blocks.getArray(e)){
			block.breakNaturally(tool.getRandom());
			
		}
	}
```
and tested with this
```vb
on right click:
	break clicked block naturally using {_null}
```
which resulted in this [ERROR](https://paste.md-5.net/hatebuceka.cs)

**Comment by Pikachu920** [comment]
are you trying to implement the behavior I'm talking about or the one Simba is talking about the?

**Comment by ShaneBeee** [comment]
> are you trying to implement the behavior I'm talking about or the one Simba is talking about the?

Im not exactly sure what you are wanting me to try?

You originally said 
```
have you tried without the conditions? calling it with null probably does the same thing as the no arg call
```
So I remove the condition and get an NPE
Im just not sure what you are wanting me to try.
I mean, i could create an item stack using air rather than null, but that makes no sense, when the condition is shorter and has the same affect

**Comment by ShaneBeee** [comment]
Are you referring to:
`
what about a case where tool is null not because the expression is null but because the value was. should that really still break the block?
`

Anyways I tested with 
```vb
break clicked block naturally using {_null}
```
and it does not throw any errors and breaks the same as if using an empty hand

**Comment by Pikachu920** [comment]
yeah, I'm just looking to get opinions on whether or not it's a good thing that it still breaks

**Comment by ShaneBeee** [comment]
> yeah, I'm just looking to get opinions on whether or not it's a good thing that it still breaks

Ah gotcha gotcha. Okay well my opinion would be yes.
Because a skripter could write a condition like
```vb
on right click:
    if player's tool is a diamond shovel:
        break clicked block naturally using player's tool
```

**Comment by Pikachu920** [comment]
but that doesn't represent the thing I'm talking about since player's tool is non-null beyond the first condition

**Comment by Snow-Pyon** [comment]
> in my mind it wouldn't break because i obviously intended it to break with _some_ tool

I already see people reporting that the effect doesn't work. I am pretty sure people expect it to break the block even if the given tool isn't set.

**Comment by Nicofisi** [comment]
Doesn't the vast majority of effects do nothing when at least one thing is not set though?

**Comment by bensku** [comment]
Fixed a few merge conflicts, and a "few" more commits got added there. Oops.

**Comment by ShaneBeee** [comment]
> Fixed a few merge conflicts, and a "few" more commits got added there. Oops.
> 
> Edit: Fixed it. Apparently Github needed a reminder to compare branches again.

Okay updated the docs to follow suit!

**Comment by ShaneBeee** [comment]
Im going to close this for a hot minute.

Im deleting my messy messy fork, re-forking and going to redo this, so i can also submit some other PRs
